crypto/crypto/src/x25519.rs

```diff
@@ -1,340 +1,151 @@
+    &self.0
+  }
 }
 
-//////////////////////
-// X25519PublicKey Traits //
-//////////////////////
-
-impl<'a> From<&'a X25519EphemeralPrivateKey> for X25519PublicKey {
-    fn from(ephemeral: &'a X25519EphemeralPrivateKey) -> X25519PublicKey {
-        X25519PublicKey(x25519_dalek::PublicKey::from(&ephemeral.0))
-    }
-}
+//
+// Traits implementations
+// ======================
+//
 
-impl<'a> From<&'a X25519StaticPrivateKey> for X25519StaticPublicKey {
-    fn from(ephemeral: &'a X25519StaticPrivateKey) -> X25519StaticPublicKey {
-        X25519StaticPublicKey(X25519PublicKey(x25519_dalek::PublicKey::from(&ephemeral.0)))
-    }
+impl From<&PrivateKeyBytes> for PublicKeyBytes {
+  fn from(private_key: &PrivateKeyBytes) -> Self {
+    private_key.public_key()
+  }
 }
 
-impl std::hash::Hash for X25519PublicKey {
-    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        let encoded_pubkey = self.0.as_bytes();
-        state.write(encoded_pubkey);
-    }
-}
+impl std::convert::TryFrom<&[u8]> for PrivateKeyBytes {
+  type Error = traits::CryptoMaterialError;
 
-impl PartialEq for X25519PublicKey {
-    fn eq(&self, other: &X25519PublicKey) -> bool {
-        *self.0.as_bytes() == *other.0.as_bytes()
-    }
+  fn try_from(private_key_bytes: &[u8]) -> Result<Self, Self::Error> {
+    let private_key_bytes: [u8; PRIVATE_KEY_SIZE] = private_key_bytes
+      .try_into()
+      .map_err(|_| traits::CryptoMaterialError::DeserializationError)?;
+    Ok(Self(private_key_bytes))
+  }
 }
 
-impl Eq for X25519PublicKey {}
-
-impl PublicKey for X25519PublicKey {
-    type PrivateKeyMaterial = X25519EphemeralPrivateKey;
-}
+impl std::convert::TryFrom<&[u8]> for PublicKeyBytes {
+  type Error = traits::CryptoMaterialError;
 
-impl PublicKey for X25519StaticPublicKey {
-    type PrivateKeyMaterial = X25519StaticPrivateKey;
+  fn try_from(public_key_bytes: &[u8]) -> Result<Self, Self::Error> {
+    let public_key_bytes: [u8; PUBLIC_KEY_SIZE] = public_key_bytes
+      .try_into()
+      .map_err(|_| traits::CryptoMaterialError::WrongLengthError)?;
+    Ok(Self(public_key_bytes))
+  }
 }
 
-impl TryFrom<&[u8]> for X25519StaticPublicKey {
-    type Error = CryptoMaterialError;
-    fn try_from(bytes: &[u8]) -> Result<X25519StaticPublicKey, CryptoMaterialError> {
-        if bytes.len() != X25519_PUBLIC_KEY_LENGTH {
-            return Err(CryptoMaterialError::DeserializationError);
-        }
-        let mut bits = [0u8; X25519_PRIVATE_KEY_LENGTH];
-        bits.copy_from_slice(&bytes[..X25519_PRIVATE_KEY_LENGTH]);
-        Ok(X25519StaticPublicKey(X25519PublicKey(
-            x25519_dalek::PublicKey::from(bits),
-        )))
-    }
+impl traits::ValidKey for PrivateKeyBytes {
+  fn to_bytes(&self) -> Vec<u8> {
+    self.0.to_vec()
+  }
 }
 
-impl Length for X25519StaticPublicKey {
-    fn length(&self) -> usize {
-        X25519_PUBLIC_KEY_LENGTH
-    }
+impl traits::ValidKey for PublicKeyBytes {
+  fn to_bytes(&self) -> Vec<u8> {
+    self.0.to_vec()
```

Comment:
ahh, why do we need as_slice if we already have this? do we call as_slice?

Reply:
`&to_bytes()` allocates memory, `as_slice()` doesn't